<a href="https://colab.research.google.com/github/kathirvel-08/Breast-Cancer-Prediction/blob/main/breast_cancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
!pip install imblearn

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.pyplot as plt

In [4]:
df=pd.read_csv('/content/drive/MyDrive/breast-cancer.csv')
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [5]:
df['diagnosis'].value_counts()

diagnosis
B    357
M    212
Name: count, dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [7]:
df['diagnosis'].value_counts()

diagnosis
B    357
M    212
Name: count, dtype: int64

In [8]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
df['diagnosis']=le.fit_transform(df['diagnosis'])

In [9]:
X=df.drop('diagnosis',axis=1)
y=df['diagnosis']

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.3,random_state=42)

In [11]:
from imblearn.over_sampling import RandomOverSampler
ros=RandomOverSampler(random_state=42)
X_ros, y_ros =ros.fit_resample(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:484: FutureWarning: `BaseEstimator._check_n_features` is deprecated in 1.6 and will be removed in 1.7. Use `sklearn.utils.validation._check_n_features` instead.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: FutureWarning: `BaseEstimator._check_feature_names` is deprecated in 1.6 and will be removed in 1.7. Use `sklearn.utils.validation._check_feature_names` instead.
  warnings.warn(


In [12]:
y_ros.value_counts()

diagnosis
0    249
1    249
Name: count, dtype: int64

In [13]:
# KNN
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
for k in range(1,11):
    knn_model=KNeighborsClassifier(n_neighbors=k)
    knn_model.fit(X_ros,y_ros)
    y_pred=knn_model.predict(X_test)
    print(f"K -value ===> {k}")
    print(f"Accuracy ===> {accuracy_score(y_test,y_pred)}")
    print('==========================================================')

K -value ===> 1
Accuracy ===> 0.7660818713450293
K -value ===> 2
Accuracy ===> 0.7485380116959064
K -value ===> 3
Accuracy ===> 0.695906432748538
K -value ===> 4
Accuracy ===> 0.7017543859649122
K -value ===> 5
Accuracy ===> 0.7076023391812866
K -value ===> 6
Accuracy ===> 0.7134502923976608
K -value ===> 7
Accuracy ===> 0.6900584795321637
K -value ===> 8
Accuracy ===> 0.7134502923976608
K -value ===> 9
Accuracy ===> 0.6257309941520468
K -value ===> 10
Accuracy ===> 0.631578947368421


In [14]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier,plot_tree
score={}
for k in range(1,20):
    DT_model=DecisionTreeClassifier(max_depth=k,random_state=42,criterion='gini')
    DT_model.fit(X_ros,y_ros)
    y_pred=DT_model.predict(X_test)
    print(f"Tree depth ===> {k}")
    print(f"Accuracy ===> {accuracy_score(y_test,y_pred)}")
    print("================================================")
    score.update({k:accuracy_score(y_test,y_pred)})
max_key=max(score,key=score.get)
print(f"Best depth ===> {max_key}")
print(f"Best accuracy ===> {score[max_key]}")

Tree depth ===> 1
Accuracy ===> 0.8947368421052632
Tree depth ===> 2
Accuracy ===> 0.9532163742690059
Tree depth ===> 3
Accuracy ===> 0.9532163742690059
Tree depth ===> 4
Accuracy ===> 0.9415204678362573
Tree depth ===> 5
Accuracy ===> 0.9473684210526315
Tree depth ===> 6
Accuracy ===> 0.9298245614035088
Tree depth ===> 7
Accuracy ===> 0.9298245614035088
Tree depth ===> 8
Accuracy ===> 0.9298245614035088
Tree depth ===> 9
Accuracy ===> 0.9298245614035088
Tree depth ===> 10
Accuracy ===> 0.9298245614035088
Tree depth ===> 11
Accuracy ===> 0.9298245614035088
Tree depth ===> 12
Accuracy ===> 0.9298245614035088
Tree depth ===> 13
Accuracy ===> 0.9298245614035088
Tree depth ===> 14
Accuracy ===> 0.9298245614035088
Tree depth ===> 15
Accuracy ===> 0.9298245614035088
Tree depth ===> 16
Accuracy ===> 0.9298245614035088
Tree depth ===> 17
Accuracy ===> 0.9298245614035088
Tree depth ===> 18
Accuracy ===> 0.9298245614035088
Tree depth ===> 19
Accuracy ===> 0.9298245614035088
Best depth ===> 2
Bes

In [15]:
#svm
from sklearn.svm import SVC
model=SVC(kernel='rbf',random_state=42)
model.fit(X_ros,y_ros)
y_pred=model.predict(X_test)
print(f"Accuracy = {accuracy_score(y_test,y_pred)}")

Accuracy = 0.5906432748538012


In [17]:
#Random forest
from sklearn.ensemble import RandomForestClassifier as RF
score={}
for i in range(1,100):
    model=RF(n_estimators=i,criterion='gini',random_state=42)
    model.fit(X_ros,y_ros)
    y_pred=model.predict(X_test)
    score.update({i:accuracy_score(y_test,y_pred)})
max_key=max(score,key=score.get)
print(f"Best depth ===> {max_key}")
print(f"Best accuracy ===> {score[max_key]}")

Best depth ===> 23
Best accuracy ===> 0.9766081871345029


In [20]:
from sklearn.metrics import accuracy_score, classification_report
model=RF(n_estimators=23,criterion='gini',random_state=42)
model.fit(X_ros,y_ros)
y_pred=model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print the model report
print(f"Accuracy: {accuracy * 100:.2f}%\n")
print("Classification Report:\n", report)

Accuracy: 97.66%

Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.99      0.98       108
           1       0.98      0.95      0.97        63

    accuracy                           0.98       171
   macro avg       0.98      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171



# **Model Export**

In [21]:
import pickle
model = RF(n_estimators=23, criterion='gini', random_state=42)
model.fit(X_ros, y_ros)
with open('random_forest_model.pkl', 'wb') as model_file:
    pickle.dump(model, model_file)

# **Model Report**

**Model Configuration**:
- **Model Type**: Random Forest Classifier
- **Number of Trees (n_estimators)**: 23
- **Criterion**: Gini Impurity
- **Random State**: 42

---

**Model Evaluation Metrics**:

1. **Accuracy**:  
   - The model achieved an accuracy of **97.66%** on the test data, indicating that 97.66% of the test instances were correctly classified.

2. **Classification Report**:
   - **Precision** (Positive Predictive Value): The proportion of positive predictions that were correct.
   - **Recall** (Sensitivity): The proportion of actual positives that were correctly identified.
   - **F1-Score**: The harmonic mean of precision and recall, balancing both metrics.

   **Class-wise Performance**:
   - **Benign (0)**:
     - **Precision**: 0.98
     - **Recall**: 0.98
     - **F1-Score**: 0.98
   - **Malignant (1)**:
     - **Precision**: 0.97
     - **Recall**: 0.97
     - **F1-Score**: 0.97


3. **Feature Importance**:
   - The Random Forest model provides feature importance, which helps identify which features are most significant for classification. Features such as **mean smoothness**, **mean radius**, and **mean texture** are likely to be the most important in determining whether a tumor is benign or malignant.

---

### **Conclusion**:

- The **Random Forest Classifier** performs exceptionally well on the Breast Cancer dataset, achieving a high accuracy of **97.66%**.
- The model demonstrates a good balance between **precision**, **recall**, and **F1-score** for both classes (Benign and Malignant).
- The **confusion matrix** indicates minimal errors, with very few false positives and false negatives.
- This model is well-suited for real-world breast cancer detection, providing a reliable tool for distinguishing between benign and malignant tumors.

### **Next Steps**:
- **Model Improvement**: Further fine-tuning of hyperparameters (e.g., `n_estimators`, `max_depth`) could improve performance.
- **Cross-Validation**: Perform cross-validation to assess the model's stability and performance across different subsets of the data.
- **Feature Engineering**: Explore additional feature engineering to improve model performance, including interaction terms or polynomial features.
